### PR TITLE
Add new API to avoid repeated expensive calls to getCredentials()

### PR DESCRIPTION
**Pull request by iambmelt**
Impetus:
- https://portal.microsofticm.com/imp/v3/incidents/details/230377348/home

`getCredentials()` is expensive: it decrypts the cache.

This PR adds new API to the cache to try and avoid making repeated expensive calls. [Existing tests](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/androidTest/java/com/microsoft/identity/common/MsalOAuth2TokenCacheTest.java) cover this API, there are no functional changes.

Inspired by the following logs, where a _lot_ of time is spent loading from the cache.... Please note that this log capture _does not_ contain the optimization shipped in #1271 -- this PR builds on that
```
2021-04-23T12:33:35.8670000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: BrokerOAuth2TokenCache:save (5 args), Message:  [2021-04-23 12:33:35 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Saving to FOCI cache? [true] Android 28
2021-04-23T12:33:36.4880000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: MsalOAuth2TokenCache:scopesIntersect, Message:  [2021-04-23 12:33:36 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Scopes intersect. Android 28
2021-04-23T12:33:36.4890000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: MsalOAuth2TokenCache:scopesIntersect, Message:  [2021-04-23 12:33:36 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] [https://substrate.office.com/search/.default] contains [https://substrate.office.com/search/.default] Android 28
2021-04-23T12:33:36.4890000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: MsalOAuth2TokenCache:deleteAccessTokensWithIntersectingScopes, Message:  [2021-04-23 12:33:36 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Removing credential: AccountCredentialBase{mAdditionalFields={}} Android 28
2021-04-23T12:33:36.4900000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: SharedPreferencesAccountCredentialCache, Message:  [2021-04-23 12:33:36 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Removing Credential... Android 28
2021-04-23T12:33:36.9420000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: SharedPreferencesFileManager, Message:  [2021-04-23 12:33:36 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Removing cache key Android 28
2021-04-23T12:33:36.9430000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: SharedPreferencesFileManager, Message:  [2021-04-23 12:33:36 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Removed cache key [foo.baz-login.windows.net-accesstoken-1fec8e78-bce4-4aaf-ab1b-5451cc387264-e0793d39-0939-496d-b129-198edd916feb-https://substrate.office.com/search/.default openid offline_access profile] Android 28
2021-04-23T12:33:36.9430000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: SharedPreferencesAccountCredentialCache, Message:  [2021-04-23 12:33:36 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Credential was removed? [true] Android 28
2021-04-23T12:33:36.9430000	WARN	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: SharedPreferencesAccountCredentialCache, Message:  [2021-04-23 12:33:36 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Deserialization failed. Skipping Credential Android 28
2021-04-23T12:33:36.9740000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: BrokerOAuth2TokenCache:updateApplicationMetadataCache, Message:  [2021-04-23 12:33:36 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Cache updated successfully? [true] Android 28
2021-04-23T12:33:39.0250000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: SharedPreferencesAccountCredentialCache, Message:  [2021-04-23 12:33:39 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Found [1] Accounts... Android 28
2021-04-23T12:33:40.4010000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: MsalOAuth2TokenCache:getAccount, Message:  [2021-04-23 12:33:40 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Found 1 accounts Android 28
2021-04-23T12:33:40.7050000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: SharedPreferencesAccountCredentialCache, Message:  [2021-04-23 12:33:40 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Found [1] Accounts... Android 28
2021-04-23T12:33:41.0590000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: SharedPreferencesAccountCredentialCache, Message:  [2021-04-23 12:33:41 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Found [1] Accounts... Android 28
2021-04-23T12:33:43.5970000	INFO	MSALCommonLoggerCallback	29479	73853	MSALCommon: Tag: AbstractBrokerController, Message:  [2021-04-23 12:33:43 - thread_name : silent-1, correlation_id : 34cb8e55-033a-4e48-9b20-2601f2a3adbc] Saving result to Android account manager Android 28
```

### CODE REVIEW

**Comment by AdamBJohnsonx** [comment]
can we put together a jmh run for this kind of thing?

**Comment by iambmelt** [comment]
> can we put together a jmh run for this kind of thing?

@AdamBJohnsonx Is this trivial to do? We're not currently consuming jmh in this project

**Review by AdamBJohnsonx — APPROVED** [review]

**Review by rpdome — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java [hunk 1]
```diff
@@ -693,6 +693,9 @@ public ICacheRecord load(@NonNull final String clientId,
                 account.getAuthorityType()
         );
 
+        // 'Preloading' our credentials to avoid repeated expensive cache hits
+        final List<Credential> allCredentials = mAccountCredentialCache.getCredentials();
+
         // Load the AccessTokens
         final List<Credential> accessTokens = mAccountCredentialCache.getCredentialsFilteredBy(
```
Comment (iamgusain) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java:700]: >mAccountCredentialCache.getCredentialsFilteredBy [](http://example.com/codeflow?start=46&length=48)

nit: we might as well create a function that can return all the tokens.. in one call.. mAccountCredentialCache.getCredentials(List<CredentialTypes>, OtherfilteringCrieteria)

Comment (iamgusain) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java:700]: I think we already have one.. 

getCredentialsFilteredBy(@Nullable final String homeAccountId,
                                                     @Nullable final String environment,
                                                     @NonNull final Set<CredentialType> credentialTypes,
                                                     @Nullable final String clientId,
                                                     @Nullable final String realm,
                                                     @Nullable final String target,
                                                     @Nullable final String authScheme,
                                                     @Nullable final String requestedClaims) {

Comment (iamgusain) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java:700]: ah.. its also internally doing the same ting in a for loop... nvm

**Review by iamgusain — APPROVED** [review]
:shipit:

**Pull request merged**